river/cluster/denstream.py

```diff
@@ -0,0 +1,491 @@
+
+        n_clusters = len(clusters)
+
+        return n_clusters, clusters
+
+    def learn_one(self, x, sample_weight=None):
+        """Incrementally trains the model.
+
+        Training a new point includes two following main tasks. First, the two weight thresholds
+        for `p-` and `o-` micro clusters will calculated. For each micro cluster, if its weight
+        is smaller than its respective threshold, that  micro cluster will be deleted.
+
+        Parameters
+        ----------
+        x
+            The point that is passed through the function to be learned by the algorithm
+
+        sample_weight
+            Weight of the sample (point)
+
+        Returns
+        -------
+        self
+
+        """
+        self.time_stamp += 1
+
+        # initialize
+        if not self.initialized:
+            mc_from_x = MicroCluster(
+                x=x,
+                timestamp=self.time_stamp,
+                decaying_factor=self.decaying_factor,
+                current_time=self.time_stamp,
+            )
+            if mc_from_x.weight >= self.tolerance_factor * self.core_weight_threshold:
+                self.p_micro_clusters[0] = mc_from_x
+            else:
+                self.o_micro_clusters[0] = mc_from_x
+            self.initialized = True
+            return self
+
+        # update current_time of all micro clusters
+        for p_micro_cluster in self.p_micro_clusters.values():
+            p_micro_cluster.current_time = self.time_stamp
+        for o_micro_cluster in self.o_micro_clusters.values():
+            o_micro_cluster.current_time = self.time_stamp
+
+        # merge x
+        self._merge(x)
+
+        # run through if conditions
+        if self.time_stamp % self._time_point == 0:
+            for i, p_micro_cluster_i in list(self.p_micro_clusters.items()):
+                if (
+                    p_micro_cluster_i.weight
+                    < self.tolerance_factor * self.core_weight_threshold
+                ):
+                    # delete c_p
+                    self.p_micro_clusters.pop(i)
+            for j, o_micro_cluster_j in list(self.o_micro_clusters.items()):
+                # calculate xi
+                xi = (
+                    2
+                    ** (
+                        -self.decaying_factor
+                        * (
+                            self.time_stamp
+                            - o_micro_cluster_j.creation_time
+                            + self._time_point
+                        )
+                    )
+                    - 1
+                ) / (2 ** (-self.decaying_factor * self._time_point) - 1)
+                if o_micro_cluster_j.weight < xi:
+                    # delete c_o
+                    self.o_micro_clusters.pop(j)
+        return self
+
+    def predict_one(self, x, sample_weight=None):
+        """Predict the cluster index for each sample
+
+        `predict_one()` is used to handle the case when a clustering request arrives.
```

Comment:
Most of this information is duplicated in the class docstring. I suggest moving the relevant (and missing) information to the class docstring and use the default docstring here.

> using the already implemented query function `_find_closest_cluster_index()`.

Avoiud this kind of comment. The method is hidden from the user so it can be confusing. On the other hand, for non-developers this information is irrelevant.

Reply:
Resolved. Also deleted the whole docstring of this function.